Create floc.bs

xyaoinum opened the pull request:
@jkarlin, @domenic: PTAL. Thanks!


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***

### :boom: Error: 400 Bad Request :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Feb 18, 2021, 3:30 PM UTC)_.

<details>
<summary>More</summary>


PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [CSS Spec Preprocessor](https://api.csswg.org/bikeshed/) - CSS Spec Preprocessor is the web service used to build Bikeshed specs.

:link: [Related URL](https://api.csswg.org/bikeshed/?url=https%3A%2F%2Fraw.githubusercontent.com%2FWICG%2Ffloc%2F9990c9d230ab505984056175e9f89e8222c75917%2Ffloc.bs&md-warning=not%20ready)

```
Error running preprocessor, returned code: 2.
FATAL ERROR: The document requires at least one &lt;pre class=metadata> block.
 ✘  Did not generate, due to fatal errors
```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20WICG/floc%2337.)._
</details>

domenic commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h4 id="adoption-phase">Adoption phase</h4>
+  To make the adoption easier, the user agent may relax the opt-in requirement when third-party cookies still exist. For example, pages with ads resources are an approximation of the pages that are going to opt-in to interest cohort computation in the long run. Thus, at the adoption phase, the page can be eligible to be included in the interest cohort computation if there are ads resources in the ...
+
+  Besides, at the adoption phase, the browser can use the existing cookie  to approximate the interest cohort permission setting. For example, when cookies are cleared, previous page visits should not be used for interest cohort computation; accessing to the interest cohort within a <a>document</a> should be denied if cookie access is not allowed in the document, or when third-party cookies are d ...
```
```suggestion
  Additionally, during the adoption phase, the browser can use existing cookie settings to approximate the interest cohort permission setting. For example, when cookies are cleared, previous page visits should not be used for interest cohort computation; accessing to the interest cohort within a {{Document}} should be denied if cookie access is not allowed in the document, or when third-party cookies are disallowed in general.
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  It should be clear that FLoC will never be able to prevent all misuse. There will be categories that are sensitive in contexts that weren't predicted. Beyond FLoC's technical means of preventing abuse, sites that use cohorts will need to ensure that people are treated fairly, just as they must with algorithmic decisions made based on any other data today.
+
+  <h3 id="tracking-people-via-their-interest-cohort">Tracking people via their interest cohort</h3>
+  A cohort could be used as a user identifier. It may not have enough bits of information to individually identify someone, but in combination with other information (such as an IP address), it might. One design mitigation is to ensure cohort sizes are large enough that they are not useful for tracking. The Privacy Budget explainer points towards another relevant tool that FLoC could be constrain ...
```
I would replace the last sentence with something more vague. Such as:

> If the user agent believes that too many high-entropy bits of information have already been consumed by a given {{Document}}, then the {{Document/interestCohort()}} algorithm allows them to return a rejected promise per its "exceptional circumstances" clause, which can help mitigate such tracking.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h3 id="tracking-people-via-their-interest-cohort">Tracking people via their interest cohort</h3>
+  A cohort could be used as a user identifier. It may not have enough bits of information to individually identify someone, but in combination with other information (such as an IP address), it might. One design mitigation is to ensure cohort sizes are large enough that they are not useful for tracking. The Privacy Budget explainer points towards another relevant tool that FLoC could be constrain ...
+
+  The interest cohorts over time can also be used as a user identifier. To mitigate the fingerprinting risk, the browser should ensure that the cohort does not update too often.
```
This does not make sense to me, as a non-domain expert. Wouldn't a stable identifier be *more* useful for as a user identifier? Probably this needs some expanding.

xyaoinum replied:
You are right. I updated this section.

domenic commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h4 id="compute-eligibility">Eligibility for a page to be included in the interest cohort computation</h4>
+  By default, a page is eligible for the interest cohort computation if the <a href="#the-api">document.interestCohort</a> API was used in the page during the page load.
+
+  The page can opt-out itself for the interest cohort computation through the interest-cohort <a>permissions policy</a> header.
```
This permissions policy needs to be defined somewhere. It's pretty simple; see examples like https://html.spec.whatwg.org/multipage/infrastructure.html#policy-controlled-features and https://xhr.spec.whatwg.org/#feature-policy-integration

xyaoinum replied:
Done.

domenic reviewed (changes requested):
Looking good!

I have three major categories of feedback:

1. Nits and convention guidance, e.g. on better use of Bikeshed
2. The suggestion to give an actual algorithm with method steps for the `interestCohort()` method, and to move some of the normative requirements there (instead of stating them in the privacy section).
3. Larger open issues around improving interoperability

We should address (1) before merging. (2) would also be ideal, but we could put it off and do that in followup PRs, especially for the trickier bits. (3) we should move to issues, although I've noted them inline here.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+</pre>
 
+<pre class="anchors">
+spec: html; urlPrefix: https://html.spec.whatwg.org/
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+    type: dfn
+        text: document; url: document
+
+spec: html; urlPrefix: https://www.w3.org/TR/permissions-policy/
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+    type: dfn
+        text: permissions policy; url: permissions-policy
+
+spec: WebIDL; urlPrefix: https://heycam.github.io/webidl/#
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+    Creating an ad based on the interest cohort:
+
+    <pre class="lang-js">
+      cohort = await document.interestCohort();
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+    <pre class="lang-js">
+      cohort = await document.interestCohort();
+      url = new URL("https://ads.example/getCreative");
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+      cohort = await document.interestCohort();
+      url = new URL("https://ads.example/getCreative");
+      url.searchParams.append("cohort", cohort);
+      creative = await fetch(url);
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+</section>
+
+<section>
+  <h2 id="interest-cohort">Interest Cohort</h2>
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+<section>
+  <h2 id="interest-cohort">Interest Cohort</h2>
+
+  The interest cohort is a user's assigned interest group under a particular <a href="#cohort-assignment-algorithm">assignment algorithm</a>. An interest cohort comprises an <a>interest cohort id</a> and a <a>version</a>.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+  The interest cohort is a user's assigned interest group under a particular <a href="#cohort-assignment-algorithm">assignment algorithm</a>. An interest cohort comprises an <a>interest cohort id</a> and a <a>version</a>.
+
+  The <dfn>interest cohort id</dfn> represents the group that the user is assigned to by the <a href="#cohort-assignment-algorithm">algorithm</a>. When exposing to the Web as a string, it’s recommended to use a decimal or hexadecimal representation (e.g “17319”, “43A7”). At the minimum, it should not contain the dot character to avoid any ambiguity in the <a>string representation</a> of interest  ...
```
Changed to decimal only. And I'm requiring it to be no more than 32 bits. But I'm not sure how to justify the choice of number of bits that's lower than 32. So may address it in a future PR.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+<section>
+  <h2 id="the-api">The API</h2>
+
+  The interest cohort API lives under the <a>document</a> object since the access permission is tied to the document scope.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+        Promise&lt;DOMString&gt; interestCohort();
+    };
+  </pre>
+
```
Done % the [=promise/Resolve=] part, got an error: No 'dfn' refs found for 'resolve' with for='['promise']'.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+  The promise will reject on permission errors. See the <a href="#access-eligibility">Permission</a> section for more details.
+
+  In exceptional situations when the <a>version</a> is unavailable (i.e. the algorithm is unknown yet), the browser could do whatever it considers reasonable. For example, it could throw an exception, or use a placeholder version, etc.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+<section>
+  <h2 id="interpretation-from-ad-techs">Interpretation from ad techs</h2>
+  Ad tech companies are supposed to observe the habits of each <a href="#string-representation">interest cohort</a> and ad targeting can then be partly based on what group the person falls into. In general, different <a href="#version">versions</a> should not be compatible. However, the browser vendors could share additional context of the version and/or could provide recommendations on the compa ...
```
Changed to dictionary.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+<section>
+  <h2 id="interpretation-from-ad-techs">Interpretation from ad techs</h2>
+  Ad tech companies are supposed to observe the habits of each <a href="#string-representation">interest cohort</a> and ad targeting can then be partly based on what group the person falls into. In general, different <a href="#version">versions</a> should not be compatible. However, the browser vendors could share additional context of the version and/or could provide recommendations on the compa ...
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+<section>
+  <h2 id="interpretation-from-ad-techs">Interpretation from ad techs</h2>
+  Ad tech companies are supposed to observe the habits of each <a href="#string-representation">interest cohort</a> and ad targeting can then be partly based on what group the person falls into. In general, different <a href="#version">versions</a> should not be compatible. However, the browser vendors could share additional context of the version and/or could provide recommendations on the compa ...
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+<section>
+  <h2 id="algorithm">Algorithm</h2>
+  <em>This section uses "cohort" as the short name for "<a>interest cohort id</a>".</em>
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <em>This section uses "cohort" as the short name for "<a>interest cohort id</a>".</em>
+
+  <h3 id="cohort-assignment-algorithm">Cohort assignment algorithm</h3>
+  The browser could use machine learning algorithms to develop the cohort.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h3 id="updating-invalidating-the-cohort">Updating/Invalidating the cohort</h3>
+  The browser should recompute the cohort on a regular basis so that the user’s latest browsing behavior can be captured. The browser should also limit the computation frequency to mitigate the <a href="#tracking-people-via-their-interest-cohort">fingerprinting risks</a>.
+
+  The browser should respect any change in user preferences (i.e. dedicated permission setting, or cookie setting at the <a href="#adoption-phase">adoption phase</a>) and may need to clear the interest cohort accordingly.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+  The browser should respect any change in user preferences (i.e. dedicated permission setting, or cookie setting at the <a href="#adoption-phase">adoption phase</a>) and may need to clear the interest cohort accordingly.
+
+  When the browsing history is deleted and the current interest cohort is derived from any deleted browsing history, the interest cohort should be cleared.
```
With the new design, the relevant discussion occurs in the last section "Recovering the browsing history from cohorts". "Cleared" should still be the right word, as the idea is to prevent updating the interest cohort too often.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+  <h3 id="permission">Permission</h3>
+  <h4 id="compute-eligibility">Eligibility for a page to be included in the interest cohort computation</h4>
+  By default, a page is eligible for the interest cohort computation if the <a href="#the-api">document.interestCohort</a> API was used in the page during the page load.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+
+  <h3 id="permission">Permission</h3>
+  <h4 id="compute-eligibility">Eligibility for a page to be included in the interest cohort computation</h4>
+  By default, a page is eligible for the interest cohort computation if the <a href="#the-api">document.interestCohort</a> API was used in the page during the page load.
```
I removed the "during the page load", as the intention is really just when "the API is used in the page".

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h4 id="compute-eligibility">Eligibility for a page to be included in the interest cohort computation</h4>
+  By default, a page is eligible for the interest cohort computation if the <a href="#the-api">document.interestCohort</a> API was used in the page during the page load.
+
+  The page can opt-out itself for the interest cohort computation through the interest-cohort <a>permissions policy</a> header.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  The user agent should offer a dedicated permission setting for the user to disallow sites from being included for interest cohort calculations.
+
+  <h4 id="access-eligibility">Permission to access the interest cohort</h4>
+  The page can restrict itself or subframes from accessing the interest cohort through the interest-cohort <a>permission policy</a> header.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  <h4 id="access-eligibility">Permission to access the interest cohort</h4>
+  The page can restrict itself or subframes from accessing the interest cohort through the interest-cohort <a>permission policy</a> header.
+
+  The user agent should offer a dedicated permission setting for the user to disallow sites from accessing the interest cohort.
```
Done.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  The user agent should offer a dedicated permission setting for the user to disallow sites from accessing the interest cohort.
+
+  <h4 id="private-browsing-mode">Private Browsing / Incognito mode</h4>
+  In private browsing / incognito mode, using <a href="#the-api">the API</a> should throw an exception.
```
Upon further check, for incognito we should probably not return a rejected promise, but should just apply whatever the algorithm is on an "empty history". PTAL at the updated one.

xyaoinum commented on floc.bs:
```diff
@@ -1 +1,149 @@
+  In private browsing / incognito mode, using <a href="#the-api">the API</a> should throw an exception.
+
+  <h4 id="adoption-phase">Adoption phase</h4>
+  To make the adoption easier, the user agent may relax the opt-in requirement when third-party cookies still exist. For example, pages with ads resources are an approximation of the pages that are going to opt-in to interest cohort computation in the long run. Thus, at the adoption phase, the page can be eligible to be included in the interest cohort computation if there are ads resources in the ...
```
Done.

xyaoinum commented:
Hi @domenic, I've addressed your comments. One other major change I made to the design is: now the baseline algorithm becomes we alway calculate on the fly, instead of calculating every X days. This is to mitigate the tracking risk. The last section "Recovering the browsing history from cohorts" discusses potentially rate limiting the calculation per-origin, to mitigate another type of privacy concern.

PTAL. Thanks!

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+
+  The <dfn>interest cohort version</dfn> identifies the <a href="#cohort-assignment-algorithm">algorithm</a> used to compute the [=interest cohort id=].
+
+  The <dfn>string representation of the interest cohort version</dfn> is browser defined. It’s recommended to make the browser vendor name as part of the version (e.g. “chrome.2.1”, “v21/mozilla”), so that when exposed to the Web, there won't be naming collisions across browser vendors.
```
```suggestion
  The <dfn>string representation of the interest cohort version</dfn> is [=implementation-defined=]. It’s recommended to make the browser vendor name as part of the version (e.g. “chrome.2.1”, “v21/mozilla”), so that when exposed to the Web, there won't be naming collisions across browser vendors.
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+
+  The <dfn>string representation of the interest cohort version</dfn> is browser defined. It’s recommended to make the browser vendor name as part of the version (e.g. “chrome.2.1”, “v21/mozilla”), so that when exposed to the Web, there won't be naming collisions across browser vendors.
+
+  <h3 id="interest-cohort-dictionary">Dictionary {{InterestCohort}}</h3>
```
I would suggest not giving this its own subsection, but instead just putting the IDL into the main API IDL block.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+<section>
+  <h2 id="the-api">The API</h2>
+
+  The interest cohort API lives under the {{document}} object since the access permission is tied to the document scope.
```
```suggestion
  The interest cohort API lives under the {{Document}} interface since the access permission is tied to the document scope.
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+
+  <pre class="idl">
+    partial interface Document {
+        Promise&lt;DOMString&gt; interestCohort();
```
```suggestion
        Promise&lt;InterestCohort&gt; interestCohort();
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  </pre>
+
+  The <dfn for="Document" method>interestCohort()</dfn> method steps are:
+  1. Let |p| be a new {{Promise}}.
```
```suggestion
  1. Let |p| be [=a new promise=].
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+
+  The <dfn for="Document" method>interestCohort()</dfn> method steps are:
+  1. Let |p| be a new {{Promise}}.
+  2. If the document is not allowed to use the "<code><a href="#interest-cohort-policy-controlled-feature">interest-cohort</a></code>" feature, reject and then return |p|.
```
```suggestion
  2. If [=this=] is not [=allowed to use=] the "<code><a href="#interest-cohort-policy-controlled-feature">interest-cohort</a></code>" feature, [=reject=] and then return |p|.
```

But, more important issue: what are you rejecting |p| with? Maybe `[=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}} and return |p|`?

Same for all other rejections. If the plan is to reject them all with the same type of error, then I'd suggest a big if statement, e.g.

```
2. If any of the following is true:
    - ...
    - ...
    - ...
  then [=reject=] |p| with ...
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  3. If the document is not allowed to access the interest cohort per user preference settings, reject and then return |p|.
+  4. If the user agent believes that too many high-entropy bits of information have already been consumed by the given document, and exposing an interest cohort would violate a budget, reject and then return |p|.
+  5. If the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a> is unavailable, reject and then return |p|.
+  6. Let |id| be [=interest cohort id=] from running the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a>.
```
This algorithm doesn't do any async work, as written. I suspect running the cohort assignment algorithm is meant to be done asynchronously, e.g. in a background thread? Or at least allowed to be?

In that case, you want to use the in parallel/queue a task structure discussed in https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors .

It might also be the case that determining some of the preconditions (e.g. user preference setting, high-entry bits, etc.) requires asynchronous work (such as contacting the browser process). In that case those would also need to move to the "in parallel" section.

xyaoinum replied:
Done. Changed to "in parallel" structure.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  The user agent should offer a dedicated permission setting for the user to disallow sites from being included for interest cohort calculations.
+
+  <h4 id="access-eligibility">Permission to access the interest cohort</h4>
+  The page can restrict itself or subframes from accessing the interest cohort through the interest-cohort [=permissions policy=] header.
```
The header doesn't work on subframes; for that you'd use `allow=""`. I'd suggest rephrasing this using the same phrasing as above, talking about the policy-controlled feature.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  <a href="#the-api">The API</a> will return a rejected promise if the user has specifically disallowed the site from accessing the [=interest cohort=].
+
+  <h4 id="private-browsing-mode">Private browsing / Incognito mode</h4>
+  The interest cohort computation <a href="#cohort-assignment-algorithm">algorithm</a> and the {{Document/interestCohort()}} API methods are applicable to the private browsing mode as well. Given that private browsing mode generally won't save history, the "information from the browsing history" is expected to just be an empty set.
```
Is this true, or it will it be a very small set (based on what they've visited in the current session)?

xyaoinum replied:
It's currently true for Chrome, but is not a guarantee / standardized. I rephrased it to: _That is, if the private browsing mode doesn't save history at all, the "information from the browsing history" is expected to just be an empty set._

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  <h4 id="adoption-phase">Adoption phase</h4>
+  To make the adoption easier, the user agent may relax the opt-in requirement while third-party cookies still exist. For example, pages with ads resources are an approximation of the pages that are going to opt-in to interest cohort computation in the long run. Thus, at the adoption phase, the page can be eligible to be included in the interest cohort computation if there are ads resources in th ...
+
+  Additionally, during the adoption phase, the browser can use the existing cookie settings to approximate the interest cohort permission setting. For example, when cookies are cleared, previous page visits should not be used for interest cohort computation; accessing to the interest cohort within a {{document}} should be denied if cookie access is not allowed in the document, or when third-party ...
```
```suggestion
  Additionally, during the adoption phase, the browser can use the existing cookie settings to approximate the interest cohort permission setting. For example, when cookies are cleared, previous page visits should not be used for interest cohort computation; accessing to the interest cohort within a {{Document}} should be denied if cookie access is not allowed in the document, or when third-party cookies are disallowed in general.
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  One design mitigation is to compute the [=interest cohort=] on the fly so that sites are less likely to observe the same [=interest cohort=] for a given user. However, one could imagine that those unstable [=interest cohorts=], when casting to a lower dimension, are still going to be stable, and the above concerns still apply though to a lesser extent.
+
+  <h3 id="recovering-the-browsing-history-from-cohorts">Recovering the browsing history from cohorts</h3>
+  Updating the [=interest cohort=] too often may let a given site learn too much about the user. An attacker could attempt to use compressed sensing to recover the browsing history.
```
What is "compressed sensing"? A hyperlink to a paper or Wikipedia article or something would be helpful.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,169 @@
+  5. If the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a> is unavailable, reject and then return |p|.
+  6. Let |id| be [=interest cohort id=] from running the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a>.
+  7. Let |version| be the [=interest cohort version=] corresponding to the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a>.
+  8. Create the {{InterestCohort}} dictionary |d|, with members {{InterestCohort/id}} being the <a href="#string-representation-of-the-interest-cohort-id">string representation</a> of |id|, and {{InterestCohort/version}} being <a href="#string-representation-of-the-interest-cohort-version">string representation</a> of |version|. Resolve |p| with |d|.
```
```suggestion
  8. Create the {{InterestCohort}} dictionary |d|, with members {{InterestCohort/id}} being the <a href="#string-representation-of-the-interest-cohort-id">string representation</a> of |id|, and {{InterestCohort/version}} being <a href="#string-representation-of-the-interest-cohort-version">string representation</a> of |version|. [=Resolve=] |p| with |d|.
```

xyaoinum replied:
Done.

domenic reviewed (commented):
Looking really good. I drilled down more on the algorithm; the big thing there is properly signaling errors, and allowing work to be done asynchronously.

I notice that most of your requirements for the algorithm or mitigations are "should" instead of "must". That might be worth a separate tracking issue to discuss. My instinct is that "should" is generally more appropriate for this kind of thing, where it's not possible to enforce via automated tests and is subject to UA policy, but others might feel differently and having a space to discuss that would be valuable.

michaelkleber commented on floc.bs:
```diff
@@ -37,113 +27,143 @@ spec: WebIDL; urlPrefix: https://heycam.github.io/webidl/#
+
+  The time series of [=interest cohorts=] can also be used as a user identifier. Sites could associate users' first-party identity with a series of [=interest cohorts=] observed over time, and could report these series to a single tracking service. The tracking service could then associate each series with the sites to know the browsing history of an individual.
+
+  One design mitigation is to compute the [=interest cohort=] on the fly so that sites are less likely to observe the same [=interest cohort=] for a given user. However, one could imagine that those unstable [=interest cohorts=], when casting to a lower dimension, are still going to be stable, and the above concerns still apply though to a lesser extent.
```
I'm not comfortable with this mitigation.  My intuition is that recalculating on the fly leads to giving out substantially _more_ information — if site X observes a person in cohort A and then shortly afterwards sees that same person in cohort B, then there is a lot of opportunity to figure out what one additional web site visit is the most likely thing to have caused this change.  Limiting the update frequency, so that each change covers a bunch of new data, seems substantially better.

xyaoinum replied:
This section talks about the "tracking" risk, and the subsequent section will talk about the mitigation for the risk you described.

Also just want to point out that the baseline algorithm in this specification is still/already on-the-fly, while there are complementing paragraphs pointing out that it's allowed to use a cached floc for performance concern and/or to use a per-site cached floc for mitigating the risk of recovering the browser history for flocs.

xyaoinum commented:
Hi @domenic: I addressed your comments. Also made some small editing/rephrasing, and under "The API" section added the restriction that the API is only available in secure context. PTAL. Thanks!

domenic commented on floc.bs:
```diff
@@ -1 +1,182 @@
+  1. Let |p| be [=a new promise=].
+  2. Run the following steps [=in parallel=]:
+
+    2.1. If any of the following is true:
```
Use just `1.` etc. for sub-bullets; otherwise this ends up being a paragraph instead of a list item.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,182 @@
+      - The user agent believes that too many high-entropy bits of information have already been consumed by the given document, and exposing an interest cohort would violate a privacy budget.
+      - The <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a> is unavailable.
+
+      then [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}} and return |p|.
```
You can't return from inside the in parallel steps. You need to queue a task to reject (but not return), which means choosing a task source. See https://html.spec.whatwg.org/multipage/webappapis.html#event-loop-for-spec-authors for more details.

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,182 @@
+
+    2.4. Create the {{InterestCohort}} dictionary |d|, with members {{InterestCohort/id}} being the <a href="#string-representation-of-the-interest-cohort-id">string representation</a> of |id|, and {{InterestCohort/version}} being <a href="#string-representation-of-the-interest-cohort-version">string representation</a> of |version|.
+
+    2.5. [=Resolve=] |p| with |d|.
```
You can't resolve the promise from an in-parallel section; you need to queue a task (and choose a task source for that queued task).

xyaoinum replied:
Done.

domenic reviewed (commented):
Almost there!

xyaoinum commented:
@domenic: I changed to "Enqueue the following steps" (referred to the way in https://github.com/privacycg/storage-access/blob/main/storage-access.bs), but I didn't find a task source suitable for this task, so I'm declaring the "interest cohort task source".  PTAL. Thanks!

xyaoinum commented:
@domenic: Thanks for the showing the way writing the algorithm. PTAL again. Thanks!

domenic commented on floc.bs:
```diff
@@ -1 +1,180 @@
+        - The <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a> is unavailable.
+
+        then:
+          1. [=Queue a task=] on the <b>|interest cohort task source|</b> to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
```
```suggestion
          1. [=Queue a task=] on the <dfn>interest cohort task source</dfn> to [=reject=] |p| with a "{{NotAllowedError}}" {{DOMException}}.
```

xyaoinum replied:
Done.

domenic commented on floc.bs:
```diff
@@ -1 +1,180 @@
+          1. Abort these steps.
+    1. Let |id| be [=interest cohort id=] from running the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a>.
+    1. Let |version| be the [=interest cohort version=] corresponding to the <a href="#cohort-assignment-algorithm">cohort assignment algorithm</a>.
+    1. [=Queue a task=] on the <b>|interest cohort task source|</b> to perform the following steps:
```
```suggestion
    1. [=Queue a task=] on the [=interest cohort task source=] to perform the following steps:
```

xyaoinum replied:
Done.

domenic reviewed (approved):
LGTM with nits on the definition/referencing for the task source. Hooray!

xyaoinum commented:
Thanks @domenic! I'll let more FLoC people take a look.

michaelkleber commented on floc.bs:
```diff
@@ -1 +1,178 @@
+
+  The <dfn>interest cohort version</dfn> identifies the <a href="#cohort-assignment-algorithm">algorithm</a> used to compute the [=interest cohort id=].
+
+  The <dfn>string representation of the interest cohort version</dfn> is [=implementation-defined=]. It’s recommended to make the browser vendor name as part of the version (e.g. “chrome.2.1”, “v21/mozilla”), so that when exposed to the Web, there won't be naming collisions across browser vendors.
```
As an exception, if two browsers choose to deliberately use the same cohort assignment algorithm, they should pick some other way to give it an unambiguous name and avoid collisions.

xyaoinum replied:
Done.

jkarlin commented on floc.bs:
```diff
@@ -1 +1,178 @@
+
+  The <dfn>interest cohort version</dfn> identifies the <a href="#cohort-assignment-algorithm">algorithm</a> used to compute the [=interest cohort id=].
+
+  The <dfn>string representation of the interest cohort version</dfn> is [=implementation-defined=]. It’s recommended to make the browser vendor name as part of the version (e.g. “chrome.2.1”, “v21/mozilla”), so that when exposed to the Web, there won't be naming collisions across browser vendors.
```
Suggest "It's recommended that the browser vendor name is part of the version"

xyaoinum replied:
Done.

jkarlin commented on floc.bs:
```diff
@@ -1 +1,178 @@
+</section>
+
+<section>
+  <h2 id="interpretation-from-ad-techs">Interpretation from ad techs</h2>
```
Suggest s/Interpretation from ad techs/Interpretation/

xyaoinum replied:
Done.

michaelkleber commented on floc.bs:
```diff
@@ -1 +1,178 @@
+  <h3 id="sensitive-information">Sensitive information</h3>
+  An [=interest cohort=] might reveal sensitive information. As a first mitigation, the browser should remove sensitive categories from its data collection. But this does not mean sensitive information can’t be leaked. Some people are sensitive to categories that others are not, and there is no globally accepted notion of sensitive categories.
+
+  Cohorts could be evaluated for fairness by measuring and limiting their deviation from population-level demographics with respect to the prevalence of sensitive categories, to prevent their use as proxies for a sensitive category. However, this evaluation would require knowing how many individual people in each cohort were in the sensitive categories, information which could be difficult or int ...
```
As an approximation, the browser could use a mechanism for recognizing which web pages are in sensitive categories.

xyaoinum replied:
Done.

jkarlin commented on floc.bs:
```diff
@@ -1 +1,178 @@
+
+<section>
+  <h2 id="interpretation-from-ad-techs">Interpretation from ad techs</h2>
+  Ad tech companies are supposed to observe the habits of each [=interest cohort=] and ad targeting can then be partly based on what group the person falls into. The browser vendors could publicly share more information about the [=interest cohort id=] (e.g. the range of numbers, whether they have semantics, etc.) or the [=interest cohort version=] (e.g. the algorithm detail, the compatibility be ...
```
Suggest s/Ad tech companies are supposed to observe/Organizations that wish to interpret cohorts can observe/

jkarlin replied:
While the API is intended for ad tech companies, the reality is that it's a public API and may be used by others. So let's replace ad tech with a more neutral term. 

e.g., s/ad techs with their modeling decisions/help with their modeling decisions/

xyaoinum replied:
Done.
